File: app/utils/LdapService.java
/**
 * Yona, 21st Century Project Hosting SW
 * <p>
 * Copyright Yona & Yobi Authors & NAVER Corp.
 * https://yona.io
 **/
package utils;

import models.User;
import models.support.LdapUser;
import org.apache.commons.lang3.StringUtils;
import play.Play;

import javax.annotation.Nonnull;
import javax.naming.Context;
import javax.naming.NamingEnumeration;
import javax.naming.NamingException;
import javax.naming.directory.*;
import java.util.Hashtable;

public class LdapService {
    public static final boolean useLdap = Play.application().configuration().getBoolean("application.use.ldap.login.supoort",false);
    private static final String HOST = Play.application().configuration().getString("ldap.host", "127.0.0.1");
    private static final String PORT = Play.application().configuration().getString("ldap.port", "389");
    private static final String BASE_DN = Play.application().configuration().getString("ldap.baseDN", "");
    private static final String DN_POSTFIX = Play.application().configuration().getString("ldap.distinguishedNamePostfix", "");
    private static final String PROTOCOL = Play.application().configuration().getString("protocol", "ldap");
    private static final String LOGIN_PROPERTY = Play.application().configuration().getString("ldap.loginProperty", "sAMAccountName");
    private static final String DISPLAY_NAME_PROPERTY = Play.application().configuration().getString("ldap.displayNameProperty", "displayName");
    private static final String USER_NAME_PROPERTY = Play.application().configuration().getString("ldap.userNameProperty", "CN");
    public static final boolean USE_EMAIL_BASE_LOGIN = Play.application().configuration().getBoolean("ldap" +
            ".options.useEmailBaseLogin", false);
    public static final boolean FALLBACK_TO_LOCAL_LOGIN = Play.application().configuration().getBoolean("ldap" +
            ".options.fallbackToLocalLogin", false);
    private static final String EMAIL_PROPERTY = Play.application().configuration().getString("ldap" +
            ".emailProperty", "mail");
    private static final String ENGLISH_NAME_PROPERTY = Play.application().configuration()
            .getString("ldap.options.englishNameAttributeName", "");
    private static final int TIMEOUT = 5000; //ms

    public LdapUser authenticate(String username, String password) throws NamingException {

        String guessedUserIdentity = guessedUser(username);

        Hashtable<String, String> env = new Hashtable<>();
        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
        env.put("com.sun.jndi.ldap.connect.timeout", ""+(TIMEOUT));
        env.put(Context.PROVIDER_URL, PROTOCOL + "://" + HOST + ":" + PORT);
        env.put(Context.SECURITY_AUTHENTICATION, "simple");
        play.Logger.info("getProperUsernameGuessing: " + getProperUsernameGuessing(guessedUserIdentity));
        env.put(Context.SECURITY_PRINCIPAL, getProperUsernameGuessing(guessedUserIdentity));
        env.put(Context.SECURITY_CREDENTIALS, password);

        DirContext authContext = new InitialDirContext(env);
        SearchResult searchResult = findUser(authContext, guessedUserIdentity, searchFilter(guessedUserIdentity));

        if (searchResult != null) {
            return getLdapUser(searchResult);
        } else {
            return null;
        }
    }

    private String guessedUser(String username) {
        if(!USE_EMAIL_BASE_LOGIN){
            return username;
        }

        String guessedUserIdentity = username;
        User user = User.findByLoginId(username);
        if (!user.isAnonymous()) {
            guessedUserIdentity = user.email;
        }

        return guessedUserIdentity;
    }

    private LdapUser getLdapUser(SearchResult searchResult) throws NamingException {
        Attributes attr = searchResult.getAttributes();
        LdapUser ldapUser = new LdapUser(attr.get(DISPLAY_NAME_PROPERTY),
                attr.get(EMAIL_PROPERTY),
                attr.get(LOGIN_PROPERTY),
                attr.get("department"));

        if(StringUtils.isNotBlank(ENGLISH_NAME_PROPERTY)){
            ldapUser.setEnglishName(attr.get(ENGLISH_NAME_PROPERTY));
        }

        return ldapUser;
    }

    private String searchFilter(@Nonnull String username) {
        if(username.contains("@")){
            return EMAIL_PROPERTY;
        } else {
            return LOGIN_PROPERTY;
        }
    }

    private String getProperUsernameGuessing(@Nonnull String username){
        if(username.contains("@")){
            return username;
        } else {
            return USER_NAME_PROPERTY + "=" + username + "," +  DN_POSTFIX;
        }
    }

    private SearchResult findUser(DirContext ctx, String username, String filter) throws NamingException {

        String searchFilter = "(" + filter + "=" + username + ")";

        play.Logger.info("filter: " + searchFilter);
        SearchControls searchControls = new SearchControls();
        searchControls.setSearchScope(SearchControls.SUBTREE_SCOPE);

        NamingEnumeration<SearchResult> results = ctx.search(BASE_DN, searchFilter, searchControls);

        SearchResult searchResult = null;
        if(results.hasMoreElements()) {
            searchResult = (SearchResult) results.nextElement();

            //make sure there is not another item available, there should be only 1 match
            if(results.hasMoreElements()) {
                System.err.println("Matched multiple users for the username: " + username);
                return null;
            }
        }

        return searchResult;
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/user/LdapService.kt
package com.github.yonaprojects.yona.domain.user

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.Hashtable
import java.util.Optional
import javax.naming.AuthenticationException
import javax.naming.CommunicationException
import javax.naming.Context
import javax.naming.NamingException
import javax.naming.directory.InitialDirContext
import javax.naming.directory.SearchControls
import javax.naming.directory.SearchResult

sealed class LdapAuthResult {
    data class Success(val user: LdapUser) : LdapAuthResult()
    object InvalidCredentials : LdapAuthResult()
    data class ConnectionFailed(val cause: Exception) : LdapAuthResult()
}

/**
 * yona의 utils/LdapService.java 대응. 실제 LDAP 서버로의 JNDI 바인딩만 담당하는
 * 얇은 글루 클래스 — 순수 로직은 LdapQueryBuilder로 분리되어 있다.
 * 2026-08-25: `LdapServiceSpec`이 Testcontainers(osixia/openldap)로 실제 LDAP 서버를 띄워
 * 실제 bind 경로까지 검증한다(InitialDirContext 생성자는 mockkConstructor로 가로챌 수 없어
 * 실제 서버가 필요했음) — search() 결과만 mockkConstructor로 시나리오별 오버라이드한다.
 */
@Component
class LdapService(
    @Value("\${yona.ldap.enabled:false}")
    val enabled: Boolean,
    @Value("\${yona.ldap.fallback-to-local-login:false}")
    val fallbackToLocalLogin: Boolean,
    @Value("\${yona.ldap.host:127.0.0.1}")
    private val host: String,
    @Value("\${yona.ldap.port:389}")
    private val port: String,
    @Value("\${yona.ldap.protocol:ldap}")
    private val protocol: String,
    @Value("\${yona.ldap.base-dn:}")
    private val baseDn: String,
    @Value("\${yona.ldap.dn-postfix:}")
    private val dnPostfix: String,
    @Value("\${yona.ldap.login-property:sAMAccountName}")
    private val loginProperty: String,
    @Value("\${yona.ldap.display-name-property:displayName}")
    private val displayNameProperty: String,
    @Value("\${yona.ldap.user-name-property:CN}")
    private val userNameProperty: String,
    @Value("\${yona.ldap.email-property:mail}")
    private val emailProperty: String,
    @Value("\${yona.ldap.department-property:department}")
    private val departmentProperty: String,
    @Value("\${yona.ldap.english-name-property:}")
    private val englishNameProperty: String,
    @Value("\${yona.ldap.use-email-base-login:false}")
    private val useEmailBaseLogin: Boolean,
    @Value("\${yona.ldap.guest-login-id-prefix:}")
    private val guestLoginIdPrefix: String,
    private val userRepository: UserRepository
) {
    private val logger = LoggerFactory.getLogger(LdapService::class.java)
    private val timeoutMs = 5000

    fun authenticate(loginIdOrEmail: String, password: String): LdapAuthResult {
        val guessedIdentity = LdapQueryBuilder.guessUser(loginIdOrEmail, useEmailBaseLogin) { candidate ->
            userRepository.findByLoginId(candidate).map { it.email }.orElse(null)
        }

        val env = Hashtable<String, String>()
        env[Context.INITIAL_CONTEXT_FACTORY] = "com.sun.jndi.ldap.LdapCtxFactory"
        env["com.sun.jndi.ldap.connect.timeout"] = timeoutMs.toString()
        env[Context.PROVIDER_URL] = "$protocol://$host:$port"
        env[Context.SECURITY_AUTHENTICATION] = "simple"
        env[Context.SECURITY_PRINCIPAL] = LdapQueryBuilder.properPrincipal(guessedIdentity, userNameProperty, dnPostfix)
        env[Context.SECURITY_CREDENTIALS] = password

        return try {
            val ctx = InitialDirContext(env)
            val searchResult = findUser(ctx, guessedIdentity)
                ?: return LdapAuthResult.InvalidCredentials

            val ldapUser = LdapQueryBuilder.parseLdapUser(
                searchResult.attributes,
                displayNameProperty, emailProperty, loginProperty, departmentProperty,
                englishNameProperty.takeIf { it.isNotBlank() }
            )
            val withGuestFlag = ldapUser.copy(
                isGuestUser = LdapQueryBuilder.isGuestUser(ldapUser.loginId, guestLoginIdPrefix)
            )
            LdapAuthResult.Success(withGuestFlag)
        } catch (e: AuthenticationException) {
            logger.warn("LDAP 인증 실패: $loginIdOrEmail")
            LdapAuthResult.InvalidCredentials
        } catch (e: CommunicationException) {
            logger.error("LDAP 서버에 연결할 수 없습니다", e)
            LdapAuthResult.ConnectionFailed(e)
        } catch (e: NamingException) {
            logger.error("LDAP 조회 중 오류가 발생했습니다", e)
            LdapAuthResult.ConnectionFailed(e)
        }
    }

    private fun findUser(ctx: InitialDirContext, identity: String): SearchResult? {
        val filterAttribute = LdapQueryBuilder.searchFilterAttribute(identity, emailProperty, loginProperty)
        val searchFilter = "($filterAttribute=$identity)"

        val searchControls = SearchControls()
        searchControls.searchScope = SearchControls.SUBTREE_SCOPE

        val results = ctx.search(baseDn, searchFilter, searchControls)
        if (!results.hasMoreElements()) {
            return null
        }
        val searchResult = results.nextElement()
        if (results.hasMoreElements()) {
            logger.warn("동일한 식별자로 여러 LDAP 사용자가 매칭되었습니다: $identity")
            return null
        }
        return searchResult
    }
}
